src/compiler/checker.ts

```diff
@@ -30818,16 +30823,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 checkNodeDeferred(memberDecl);
             }
 
-            if (computedNameType && !(computedNameType.flags & TypeFlags.StringOrNumberLiteralOrUnique)) {
-                if (isTypeAssignableTo(computedNameType, stringNumberSymbolType)) {
+            if (computedNameType && !isTypeUsableAsPropertyName(computedNameType)) {
+                if (isPatternLiteralType(computedNameType)) {
+                    indexKeyTypes.add(computedNameType);

```

Comment:
This probably warrants the "breaking change" tag, since previously these'd make a `string` index signature get added, but now you get a more specific index signature. Which is, I believe, the goal of the change, it's just... maybe breaky? It's unclear.

Reply:
Also there's the whole union thing - but the union thing can be a problem for another day.